### PR TITLE
Documentation to obtain, and configure the library

### LINKED ISSUE #7: Add documentation
The PR #3 doesn't contain documentation. I'll add it afterwards to resolve this isse.

### PR DESCRIPTION
**Pull request by marco-bertschi**
Fixes #7

### CODE REVIEW

File: docs/input/docs/installation/configure.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+```csharp
+var resolver = new YourResolver(); // Your implementation of the IIocResolver
+IocHandler.Instance.IocResolver = resolver;
+```
+
+Afterwards make sure that the dependencies from the [BBT.StrategyPattern] are registered.
+See [BBT.StrategyPattern usage with IoC] for instructions.
+
+In order for the library to work correctly it needs to register some types into the IoC container
+itself. As the implementations are private to allow for a concise API it needs to do this by itself within
```
Comment (ghost) [inline docs/input/docs/installation/configure.md]: "to allow for a concise" -> to allow a concise

**Review by ghost — CHANGES_REQUESTED** [review]
see comment in file .

File: docs/input/docs/installation/configure.md (hunk 1 above)
Comment (marco-bertschi) [reply docs/input/docs/installation/configure.md]: Thanks, adjusted

**Review by ghost — APPROVED** [review]

**Comment by marco-bertschi** [comment]
@pascalberger @georgesgoetz 
I've had another round with the documentation and added more technical principles. I think the documentation is now at a point where this PR allows us to close #7. Can you please have another look?

File: docs/input/docs/installation/configure.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+var resolver = new YourResolver(); // Your implementation of the IIocResolver
+IocHandler.Instance.IocResolver = resolver;
+```
+
+Afterwards make sure that the dependencies from the [BBT.StrategyPattern] are registered.
+See [BBT.StrategyPattern usage with IoC] for instructions.
+
+In order for the library to work correctly it needs to register some types into the IoC container
+itself. As the implementations are private to allow a concise API it needs to do this by itself within
+the `IocHandler` (see also [TestIoContainer] where this initialization is made for testing with NInject)
```
Comment (pascalberger) [inline docs/input/docs/installation/configure.md]: Replace `NInject` with `Ninject`

File: docs/input/docs/installation/configure.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
+---
+Order: 20
+Title: Configuration
+Description: Configure BBT.StructureTools
+---
+
+The BBT Structure Tools are entirely configurable by code. This section covers the
+correct IoC container registration of the infrastructure components.
```
Comment (pascalberger) [inline docs/input/docs/installation/configure.md]: Is IoC container a requirements for using this library? We should make this clear and use a clear language regarding this.

File: docs/input/docs/installation/configure.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+## Preconditions
+
+* An IoC container can be used to resolve registrations as well as internally by the library.
```
Comment (pascalberger) [inline docs/input/docs/installation/configure.md]: I don't understand the last part of this sentence (`...as well as internally be the library`).

File: docs/input/docs/installation/obtain.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+---
```
Comment (pascalberger) [inline docs/input/docs/installation/obtain.md]: I would move this under the Getting Started section

File: docs/input/docs/installation/configure.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
+---
```
Comment (pascalberger) [inline docs/input/docs/installation/configure.md]: I would move this under the Getting Started section

File: docs/input/docs/installation/obtain.md (hunk 5 above)
Comment (marco-bertschi) [reply docs/input/docs/installation/obtain.md]: Moved

File: docs/input/docs/installation/configure.md (hunk 6 above)
Comment (marco-bertschi) [reply docs/input/docs/installation/configure.md]: Moved

File: docs/input/docs/installation/configure.md (hunk 4 above)
Comment (marco-bertschi) [reply docs/input/docs/installation/configure.md]: Sentence rewritten

File: docs/input/docs/installation/configure.md (hunk 3 above)
Comment (marco-bertschi) [reply docs/input/docs/installation/configure.md]: Yes, it is - I've rewritten the sentence under the _Preconditions_ chapter to try to make this clear.

File: docs/input/docs/installation/configure.md (hunk 2 above)
Comment (marco-bertschi) [reply docs/input/docs/installation/configure.md]: Done

**Review by marco-bertschi — COMMENTED** [review]
@pascalberger I've followed your points, ready for round two

**Comment by pascalberger** [comment]
Please split into separate PRs for #7 and #21. #7 can go into `develop`, #21 needs to go into `release/1.0.0` branch.

**Comment by marco-bertschi** [comment]
> 
> 
> Please split into separate PRs for #7 and #21. #7 can go into `develop`, #21 needs to go into `release/1.0.0` branch.

Done

File: docs/input/docs/getting-started/configure.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+private void DoRegistration(Type impl, Type abstraction)
+{
+    // Register the implementation for the abstraction
+    // within your IoC container here.
+}
+```
+
+[BBT.StrategyPattern]: https://bbtsoftware.github.io/BBT.StrategyPattern/
+[BBT.StrategyPattern usage with IoC]: https://bbtsoftware.github.io/BBT.StrategyPattern/docs/usage/use-without-ioc
```
Comment (pascalberger) [inline docs/input/docs/getting-started/configure.md]: This links to the usage instructions without IoC containers

File: docs/input/docs/getting-started/principles.md [hunk 8]
```diff
@@ -20,13 +20,84 @@ according to the user-made registrations.
...
-Basically, one registration is needed for each data class type for a specific intention. The registrations
+One registration is needed for each data class type for a specific intention. The registrations
 can be seen as walking through the tree and deciding for each node which properties and child / parent objects need
 to be handled in a certain way which can be done by using what is generally called a _sub operation_. These
 _sub operations_ are specific to each of the operation, though some can be mixed (for example, while converting an object implementing
 the same interface as the master data it is being converted from the interface's properties can be copied using a sub copy).
+
+There is one registration interface per each operation supported by the library. All of these registration
+interfaces share the same raison d'être and work as containers which hold the definition for each operation and
+can retrieved from the IoC container.
```
Comment (pascalberger) [inline docs/input/docs/getting-started/principles.md]: Something is missing is this sentence (`can be retrieved`?)

File: docs/input/docs/getting-started/principles.md [hunk 9]
```diff
@@ -20,13 +20,84 @@ according to the user-made registrations.
...
+
+| Operation     | Technical                                                     |
+| ------------- |--------------------------------------------------------------:|
+| Comparison    | `ICompareRegistrations<TModelToCompare, TCompareIntention>`   |
+| Conversion    | `IConvertRegistrations<TSource, TTarget, TConvertIntention>`  |
+| Copy          | `ICopyRegistrations<T>`                                       |
+
+### Note: Registration interfaces
```
Comment (pascalberger) [inline docs/input/docs/getting-started/principles.md]: Format this as an annotation block, for example (see [Bootstrap alerts](https://getbootstrap.com/docs/4.0/components/alerts/) for possible alert classes):

```
:::{.alert .alert-info}
Consider these steps while implementing registration interfaces:

* Register your implementation of the registration in the IoC container.
* Do not traverse over multiple tree leafs when registering properties, except for reference level sub operations.
* It's usually better to unit test each registration, not the entire tree being copied.
:::
```

File: docs/input/docs/getting-started/principles.md [hunk 10]
```diff
@@ -20,13 +20,84 @@ according to the user-made registrations.
...
+
+## Controlling object initialization
+
+The copy and convert registrations can create a new instance of a target object if neccessary.
+How an object is initialized is defined by a registration of an `IInstanceCreator`
+implementation into the IoC container.
+The Copy and Convert infrastructure resolves the correct instance creator when it needs
+to create a new object while copying or converting an object.
+
+### Note: Object initialization
```
Comment (pascalberger) [inline docs/input/docs/getting-started/principles.md]: Format this as an annotation block

File: docs/input/docs/getting-started/principles.md [hunk 11]
```diff
@@ -20,13 +20,84 @@ according to the user-made registrations.
...
+default generic `InstanceCreator` because the default constructor is used there.
+
+## Processings and Interceptions
+
+Copy, Convert, and Compare can be controlled depending on the objects being processed by using
+post-, preprocessings, or interceptors.
+Please note that these operations are not taken from the IoC but passed into the operation
+as additional argument.
+
+***If possible, use the generic implementations supplied with the libary.***
```
Comment (pascalberger) [inline docs/input/docs/getting-started/principles.md]: Format this as an annotation block

**Review by pascalberger — CHANGES_REQUESTED** [review]
Comments for some smaller issues added

**Comment by marco-bertschi** [comment]
> 
> 
> Comments for some smaller issues added

Thanks for the heads-up, I did the corrections.

**Comment by pascalberger** [comment]
@marco-bertschi You didn't push your fixes

**Comment by marco-bertschi** [comment]
> 
> 
> @marco-bertschi You didn't push your fixes

I did, but the push failed. Corrected that now.

**Review by pascalberger — APPROVED** [review]

**Pull request merged**